lib/infrastructure_collector.rb

```diff
@@ -74,38 +76,112 @@ def run
+      @local_inventory.save
     end
-
-    @local_inventory.save
-
     logger.debug 'Generating vSphere session activity with currentTime request'
     VSphere.wrapped_vsphere_request { VSphere.session.serviceInstance.CurrentTime }
   end
 
   private
 
+  def updated_infrastructure(properties)
+    infrastructure = Infrastructure.where(platform_id: properties[:platform_id])
+    if infrastructure.present?
+      current_infrastructure = infrastructure.first
+      update_fields_for(current_infrastructure,properties)
+    else
+      Infrastructure.new(properties)
+    end
+  end
+
+  def update_fields_for(current_infrastructure,properties)
+    BASIC_INFRASTRUCTURE_FIELDS.each{|field| current_infrastructure[field] = properties[field] }
+    current_infrastructure[:networks] = properties[:networks] ? build_updated_networks(properties) : []
+    current_infrastructure[:volumes] = properties[:volumes] ? build_updated_volumes(properties) : []
+    current_infrastructure.record_status = 'updated'
+    current_infrastructure = update_hosts(properties,current_infrastructure)
+    current_infrastructure.save
+    current_infrastructure
+  end
+
+  def update_hosts(properties,current_infrastructure)
+    properties[:hosts].each do |h|
+      host = current_infrastructure.hosts.where(platform_id: h[:platform_id]).first
+      host.present? ? update_host(host,h) : current_infrastructure.hosts << h
+    end
+    if properties[:hosts].empty?
+      current_infrastructure.hosts = []
+    else
+      current_infrastructure.hosts = current_infrastructure.hosts.map do |h|
+        h if (properties[:hosts].map &:platform_id).include?h[:platform_id]
+      end
+    end
+    current_infrastructure
+  end
+
+  def build_updated_networks(properties)
+    properties[:networks].map do |n|
+      {'_id': n['_id'], 'kind' => n['kind'], 'name'=> n['name']}
+    end
+  end
+
+  def build_updated_volumes(properties)
+    properties[:volumes].map do |v|
+      { '_id'=> v['_id'], 'storage_bytes' => v['storage_bytes'],
+        'free_space'=> v['free_space'], 'accessible' => v['accessible'],
+        'volume_type' => v['volume_type'],'name'=> v['name']
+      }
+    end
+  end
+
+  def update_host(host,new_host)
+    host = update_basic_host_fields_for(host,new_host)
+    platform_id_number = host[:platform_id].match(/[0-9]+/)
+    host[:host_bus_adapters] = @host_bus_adapters["storageSystem-#{platform_id_number[0]}"]
+    host[:nics] = @nics["networkSystem-#{platform_id_number[0]}"]
+    host.save
+  end
+
+  def update_basic_host_fields_for(host,new_host)
+    BASIC_HOST_FIELDS.each do |field|
+      if field == :cluster_platform_id && @clusters[host[:cluster_platform_id]]
+        host[field]= @clusters[host[:cluster_platform_id]][:platform_id]
+      elsif !@clusters[host[:cluster_platform_id]] && field == :cluster_platform_id
+        host[field] = nil
+        host[:cluster] = nil
+      else
+        host[field] = new_host[field] if new_host[field].present?
```

Comment:
Maybe this could be transformed to another "elsif" (instead of an additional if inside the else)